### PR TITLE
Prevent BaseHTTPMiddleware from hiding errors of StreamingResponse

### LINKED ISSUE #1433: Raising Exceptions in sub-applications routes
### Checklist

- [X] The bug is reproducible against the latest release or `master`.
- [X] There are no similar issues or pull requests to fix it yet.

### Describe the bug

Let's start with this PR: #1262

It's about preventing raise `anyio.ExceptionGroup` in views under a `BaseHTTPMiddleware`. PR resolve that problem with nonlocal variable that stores our exception. But in the case of sub-applications, it does not work. 

As I can see (fyi I am not good at asyncio), in the case below, we reach and read a response before we raise an exception and store it to our nonlocal variable:

fragment of `BaseHTTPMiddleware.__call__`
```python
async def call_next(request: Request) -> Response:
    app_exc: typing.Optional[Exception] = None
    send_stream, recv_stream = anyio.create_memory_object_stream()

    async def coro() -> None:
        nonlocal app_exc

        async with send_stream:
            try:
                task = await self.app(scope, request.receive, send_stream.send)
            except Exception as exc:
                app_exc = exc

    task_group.start_soon(coro)

    try:
        message = await recv_stream.receive()
    except anyio.EndOfStream:
        if app_exc is not None:
            raise app_exc
        raise RuntimeError("No response returned.")
    
    ...
    response = StreamingResponse(
        status_code=message["status"], content=body_stream()
    )
    response.raw_headers = message["headers"]
    return response
```

in this moment: `except anyio.EndOfStream:` exception still no raised.

### Steps to reproduce the bug

```python
import httpx
import pytest
from fastapi import FastAPI, APIRouter
from starlette.middleware import Middleware
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
from starlette.requests import Request
from starlette.responses import Response
from starlette.routing import Route


class SomeError(Exception):
    pass


class SomeMiddleware(BaseHTTPMiddleware):
    async def dispatch(
        self, request: Request, call_next: RequestResponseEndpoint
    ) -> Response:
        return await call_next(request)

# Drop (or use not BaseHTTPMiddleware based) middleware and test works fine
app = FastAPI(middleware=[Middleware(SomeMiddleware), ])


async def simple_route(request: Request):
    raise SomeError


another_router = APIRouter(
    routes=[Route('/simple-route/', simple_route, methods=['GET'])]
)
sub_app = FastAPI()
sub_app.include_router(another_router)
app.router.mount(f'/api', sub_app)


@pytest.mark.asyncio
async def test_simple_route():
    async with httpx.AsyncClient(app=app) as client:
        with pytest.raises(SomeError):
            await client.get("http://testserver/api/simple-route/")

```

### Expected behavior

An exception was raised and caught by pytest exception

### Actual behavior

An exception wasn't raised

### Debugging material

_No response_

### Environment

macOS Monterey 12.0.1, starlette 0.17.1, Python 3.9.9

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by o-fedorov**
The issue is observed when:

*  an application uses a BaseHTTPMiddleware subclass;
*  it provides StreamingResponse;
*  there is an exception raised inside the generator wrapped with StreamingResponse.

Prior to v0.17.0 the exception was raised and the traceback was printed to the server logs. Now it is silently ignored.

Edit by @Kludex : Closes #1433

### CODE REVIEW

File: starlette/middleware/base.py [hunk 1]
```diff
@@ -47,11 +47,16 @@ async def coro() -> None:
             assert message["type"] == "http.response.start"
 
             async def body_stream() -> typing.AsyncGenerator[bytes, None]:
+                nonlocal app_exc
```
Comment (Kludex) [inline starlette/middleware/base.py]: This shouldn't be necessary, right?

**Review by Kludex — COMMENTED** [review]
Alternative to #1452

Would you mind adding this test?
```python
def test_exception_on_mounted_apps(test_client_factory):
    sub_app = Starlette(routes=[Route("/", exc)])
    app.mount("/sub", sub_app)

    client = test_client_factory(app)
    with pytest.raises(Exception) as ctx:
        client.get("/sub/")
    assert str(ctx.value) == "Exc"
```

File: starlette/middleware/base.py (hunk 1 above)
Comment (o-fedorov) [reply starlette/middleware/base.py]: You are right, there is no assignment to `app_exc`. Removed.

File: tests/middleware/test_base.py [hunk 2]
```diff
@@ -158,3 +172,12 @@ async def dispatch(self, request, call_next):
     client = test_client_factory(app)
     response = client.get("/does_not_exist")
     assert response.text == "Custom"
+
```
Comment (Kludex) [inline tests/middleware/test_base.py:175]:
```suggestion


```

Comment (o-fedorov) [reply tests/middleware/test_base.py:175]: Thank you, already did it locally

**Comment by o-fedorov** [comment]
> Alternative to #1452
> 
> Would you mind adding this test?
> 
> ```python
> def test_exception_on_mounted_apps(test_client_factory):
>     sub_app = Starlette(routes=[Route("/", exc)])
>     app.mount("/sub", sub_app)
> 
>     client = test_client_factory(app)
>     with pytest.raises(Exception) as ctx:
>         client.get("/sub/")
>     assert str(ctx.value) == "Exc"
> ```

The test is added

**Comment by Kludex** [comment]
I don't think this work if the `Response` object from the middleware is ignored, as below:

```python
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.responses import Response

class CustomHeaderMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request, call_next):
        await call_next(request)
        return Response("Do you see an exception now?")

middleware = [
    Middleware(CustomHeaderMiddleware)
]

app = Starlette(routes=routes, middleware=middleware)
```
Can you confirm? Using the same `StreamingResponse` endpoint, you created on your test with this middleware.

#1452 should be able to solve this.

**Comment by o-fedorov** [comment]
Sorry, both this approach and #1452 seems to not solving the issue  if the Response object from the middleware is ignored

**Review by Kludex — APPROVED** [review]
> Sorry, both this approach and #1452 seems to not solving the issue if the Response object from the middleware is ignored

My bad... That exception is not even raised! :laughing:

In any case, I like this PR more than the one I created because it brings the _exception conditional_ closer to where it can be retrieved.

**Comment by Kludex** [comment]
More context about this PR can be found on #1452 and #1433.

The issue here is that we check if the endpoint (or app) raised an exception only if `http.response.start` is not sent, but `StreamingResponse` (and mounted apps) pass that conditional, and we can have an exception after that point.

**Comment by Kludex** [comment]
Thanks for the PR @o-fedorov ! :)

**Pull request merged**